utils/logalloc.cc

```diff
@@ -2644,19 +2667,25 @@ region_group::del(region_group* child) {
 }
 
 void
-region_group::add(region_impl* child) {
+region_group::add(region* child_r) {
+    auto child = static_cast<region_impl*>(child_r->_impl.get());
     child->_heap_handle = _regions.push(child);
     region_group_binomial_group_sanity_check(_regions);
     update(child->occupancy().total_space());
 }
 
 void
-region_group::del(region_impl* child) {
+region_group::del(region* child_r) {
+    auto child = static_cast<region_impl*>(child_r->_impl.get());
     _regions.erase(child->_heap_handle);
     region_group_binomial_group_sanity_check(_regions);
     update(-child->occupancy().total_space());
 }
 
+void
+region_group::moved(region* old_address, region* new_address) {
```

Comment:
@avikivity Was this intentionally left empty?
How does this work with `region::operator=(region&& other)`?

Reply:
@avikivity we talked about this on the daily call

Reply:
For the region move constructor, I think the assumption here is that it is called when moving the `size_tracked_region` that wraps it, moving `size_tracked_region::_heap_handle` and the boost::heap takes care of updating the pointer to the object.

But for the move-assign operator, we explicitly call `_listener->del` on the current, moved-to _impl, and then call `_listener->moved` on the moved from `_impl`, rather than `_listener->add`.

Reply:
The reason it works is that region_group holds regions via an intrusive container. In the final code, we have

```c++
using region_heap = boost::heap::binomial_heap<size_tracked_region*,
        boost::heap::compare<region_evictable_occupancy_ascending_less_comparator>,
        boost::heap::allocator<std::allocator<size_tracked_region*>>,
        //constant_time_size<true> causes corruption with boost < 1.60
        boost::heap::constant_time_size<false>>;

class size_tracked_region : public logalloc::region {
public:
    region_heap::handle_type _heap_handle;
};
```

The heap handle has its own move constructor and reacts to moving by updating the links.

However, region doesn't know if the region_listener implementation tracks it via an intrusive container or a regular container, so it has to issue a notification.

Reply:
Please see https://github.com/scylladb/scylla/pull/11141